Enable HTTP compression for REST client only

### Discussed in https://github.com/quarkusio/quarkus/discussions/46414

<div type='discussions-op-text'>

<sup>Originally posted by **jdussouillez** February 21, 2025</sup>
I have a REST API that acts like a middleware between clients and another "external" REST API.

To send requests to the external API, I use a REST client. I want to enable HTTP compression in the REST client, but **not** in my REST API. Mostly because I have a nginx in front of my REST API that handles compression so it's handled outside of the Quarkus app.

Is this possible?

From what I saw in the documentation, to enable HTTP compression in REST client, I have to use `quarkus.http.enable-compression=true`.

https://quarkus.io/guides/rest-client#receiving-compressed-messages

> REST Client also supports receiving compressed messages using GZIP. You can enable the HTTP compression support by adding the property quarkus.http.enable-compression=true. 

But this option will also enable global HTTP compression for my API: https://quarkus.io/guides/http-reference#http-compression

I might find some way to make what I want, but I'm wondering if there is a way to easily enable compression only for REST client.</div>

> FWIW, when I build Quarkus locally I use

It's not my day I think.

```
[ERROR] Failed to execute goal net.revelc.code:impsort-maven-plugin:1.7.0:sort (sort-imports) on project quarkus-enforcer-rules: Execution sort-imports of goal net.revelc.code:impsort-maven-plugin:1.7.0:sort failed: A required class was missing while executing net.revelc.code:impsort-maven-plugin:1.7.0:sort: null
[ERROR] -----------------------------------------------------
[ERROR] realm =    plugin>net.revelc.code:impsort-maven-plugin:1.7.0
[ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
[ERROR] urls[0] = file:/home/junior/.m2/repository/net/revelc/code/impsort-maven-plugin/1.7.0/impsort-maven-plugin-1.7.0.jar
[ERROR] urls[1] = file:/home/junior/.m2/repository/com/github/javaparser/javaparser-core/3.24.2/javaparser-core-3.24.2.jar
[ERROR] urls[2] = file:/home/junior/.m2/repository/com/google/guava/guava/31.1-jre/guava-31.1-jre.jar
[ERROR] urls[3] = file:/home/junior/.m2/repository/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar
[ERROR] urls[4] = file:/home/junior/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar
[ERROR] urls[5] = file:/home/junior/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar
[ERROR] urls[6] = file:/home/junior/.m2/repository/org/checkerframework/checker-qual/3.12.0/checker-qual-3.12.0.jar
[ERROR] urls[7] = file:/home/junior/.m2/repository/com/google/errorprone/error_prone_annotations/2.11.0/error_prone_annotations-2.11.0.jar
[ERROR] urls[8] = file:/home/junior/.m2/repository/com/google/j2objc/j2objc-annotations/1.3/j2objc-annotations-1.3.jar
[ERROR] Number of foreign imports: 1
[ERROR] import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
[ERROR] 
[ERROR] -----------------------------------------------------
[ERROR] : NoClassDefFoundError: org/codehaus/plexus/util/DirectoryScanner: org.codehaus.plexus.util.DirectoryScanner
```

--- a/docs/src/main/asciidoc/rest-client.adoc
+++ b/docs/src/main/asciidoc/rest-client.adoc
@@ -1755,8 +1755,21 @@ The code uses the following pieces:
 As previously mentioned, the body parameter needs to be properly crafted by the application code to conform to the service's requirements.
 
 === Receiving compressed messages
-REST Client also supports receiving compressed messages using GZIP. You can enable the HTTP compression support by adding the property `quarkus.http.enable-compression=true`.
+REST Client also supports receiving compressed messages using GZIP and can be enabled via configuration.
 When this feature is enabled and a server returns a response that includes the header `Content-Encoding: gzip`, REST Client will automatically decode the content and proceed with the message handling.
+
+An example configuration could be:
+
+[source,properties]
+----
+# global configuration is used for all clients
+quarkus.rest-client.enable-compression=true
+
+# per-client configuration overrides the global settings for a specific client
+quarkus.rest-client.my-client.enable-compression=true
+----
+
+The REST Client falls back onto the Quarkus wide `quarkus.http.enable-compression` configuration property (which defaults to `false`) if no REST Client specific property is set.
 
 == Proxy support
 REST Client supports sending requests through a proxy.

--- a/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
+++ b/extensions/resteasy-classic/rest-client-config/runtime/src/main/java/io/quarkus/restclient/config/RestClientsConfig.java
@@ -282,6 +282,17 @@ public interface RestClientsConfig {
     Optional<MemorySize> maxChunkSize();
 
     /**
+     * Supports receiving compressed messages using GZIP.
+     * When this feature is enabled and a server returns a response that includes the header {@code Content-Encoding: gzip},
+     * REST Client will automatically decode the content and proceed with the message handling.
+     * <p>
+     * This property is not applicable to the RESTEasy Client.
+     * <p>
+     * Can be overwritten by client-specific settings.
+     */
+    Optional<Boolean> enableCompression();
+
+    /**
      * If the Application-Layer Protocol Negotiation is enabled, the client will negotiate which protocol to use over the
      * protocols exposed by the server. By default, it will try to use HTTP/2 first and if it's not enabled, it will
      * use HTTP/1.1.
@@ -599,6 +610,15 @@ public interface RestClientsConfig {
          */
         @ConfigDocDefault("8K")
         Optional<MemorySize> maxChunkSize();
+
+        /**
+         * Supports receiving compressed messages using GZIP.
+         * When this feature is enabled and a server returns a response that includes the header {@code Content-Encoding: gzip},
+         * REST Client will automatically decode the content and proceed with the message handling.
+         * <p>
+         * This property is not applicable to the RESTEasy Client.
+         */
+        Optional<Boolean> enableCompression();
 
         /**
          * If the Application-Layer Protocol Negotiation is enabled, the client will negotiate which protocol to use over the

--- a/extensions/resteasy-reactive/rest-client-jackson/deployment/src/test/java/io/quarkus/rest/client/reactive/jackson/test/ClientUsingGzipCompressionGlobalSettingTest.java
+++ b/extensions/resteasy-reactive/rest-client-jackson/deployment/src/test/java/io/quarkus/rest/client/reactive/jackson/test/ClientUsingGzipCompressionGlobalSettingTest.java
@@ -15,12 +15,13 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import io.quarkus.test.QuarkusUnitTest;
 import io.quarkus.vertx.http.Compressed;
 
-public class ClientUsingGzipCompressionTest {
+public class ClientUsingGzipCompressionGlobalSettingTest {
     @RegisterExtension
     static final QuarkusUnitTest config = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
                     .addClasses(MyResource.class, Message.class, MyClient.class))
-            .withConfigurationResource("client-using-gzip-application.properties");
+            .overrideConfigKey("quarkus.http.enable-compression", "true")
+            .overrideRuntimeConfigKey("quarkus.rest-client.my-client.url", "http://localhost:${quarkus.http.test-port:8081}");
 
     @RestClient
     MyClient client;

--- a/extensions/resteasy-reactive/rest-client-jackson/deployment/src/test/resources/client-using-gzip-application.properties
+++ b/extensions/resteasy-reactive/rest-client-jackson/deployment/src/test/resources/client-using-gzip-application.properties
@@ -1,3 +1,0 @@
-quarkus.http.enable-compression=true
-
-quarkus.rest-client.my-client.url=http://localhost:${quarkus.http.test-port:8081}

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/QuarkusRestClientBuilder.java
@@ -305,6 +305,13 @@ public interface QuarkusRestClientBuilder extends Configurable<QuarkusRestClient
     QuarkusRestClientBuilder disableDefaultMapper(Boolean disable);
 
     /**
+     * Supports receiving compressed messages using GZIP.
+     * When this feature is enabled and a server returns a response that includes the header {@code Content-Encoding: gzip},
+     * REST Client will automatically decode the content and proceed with the message handling.
+     */
+    QuarkusRestClientBuilder enableCompression(boolean enableCompression);
+
+    /**
      * Based on the configured QuarkusRestClientBuilder, creates a new instance of the given REST interface to invoke API calls
      * against.
      *

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/QuarkusRestClientBuilderImpl.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/QuarkusRestClientBuilderImpl.java
@@ -269,6 +269,12 @@ public class QuarkusRestClientBuilderImpl implements QuarkusRestClientBuilder {
     }
 
     @Override
+    public QuarkusRestClientBuilder enableCompression(boolean enableCompression) {
+        proxy.enableCompression(enableCompression);
+        return this;
+    }
+
+    @Override
     public <T> T build(Class<T> clazz) throws IllegalStateException, RestClientDefinitionException {
         return proxy.build(clazz);
     }

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientBuilderImpl.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientBuilderImpl.java
@@ -89,6 +89,7 @@ public class RestClientBuilderImpl implements RestClientBuilder {
     private Boolean trustAll;
     private String userAgent;
     private Boolean disableDefaultMapper;
+    private Boolean enableCompression;
 
     @Override
     public RestClientBuilderImpl baseUrl(URL url) {
@@ -263,6 +264,11 @@ public class RestClientBuilderImpl implements RestClientBuilder {
 
     public RestClientBuilderImpl disableDefaultMapper(Boolean disableDefaultMapper) {
         this.disableDefaultMapper = disableDefaultMapper;
+        return this;
+    }
+
+    public RestClientBuilderImpl enableCompression(boolean enableCompression) {
+        this.enableCompression = enableCompression;
         return this;
     }
 
@@ -533,10 +539,20 @@ public class RestClientBuilderImpl implements RestClientBuilder {
             clientBuilder.alpn(restClients.alpn().get());
         }
 
-        Boolean enableCompression = ConfigProvider.getConfig()
-                .getOptionalValue(ENABLE_COMPRESSION, Boolean.class).orElse(false);
-        if (enableCompression) {
-            clientBuilder.enableCompression();
+        Boolean effectiveEnableCompression = enableCompression;
+        if (effectiveEnableCompression == null) {
+            if (restClients.enableCompression().isPresent()) {
+                effectiveEnableCompression = restClients.enableCompression().get();
+            }
+        }
+        if (effectiveEnableCompression == null) {
+            var maybeGlobalEnableCompression = ConfigProvider.getConfig().getOptionalValue(ENABLE_COMPRESSION, Boolean.class);
+            if (maybeGlobalEnableCompression.isPresent()) {
+                effectiveEnableCompression = maybeGlobalEnableCompression.get();
+            }
+        }
+        if (effectiveEnableCompression != null) {
+            clientBuilder.enableCompression(effectiveEnableCompression);
         }
 
         if (proxyHost != null) {

--- a/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilder.java
+++ b/extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/RestClientCDIDelegateBuilder.java
@@ -143,6 +143,11 @@ public class RestClientCDIDelegateBuilder<T> {
                 configRoot.multipart().maxChunkSize());
         builder.property(QuarkusRestClientProperties.MAX_CHUNK_SIZE, maxChunkSize.orElse(DEFAULT_MAX_CHUNK_SIZE));
 
+        Optional<Boolean> enableCompressions = oneOf(restClientConfig.enableCompression(), configRoot.enableCompression());
+        if (enableCompressions.isPresent()) {
+            builder.enableCompression(enableCompressions.get());
+        }
+
         Boolean http2 = oneOf(restClientConfig.http2()).orElse(configRoot.http2());
         builder.property(QuarkusRestClientProperties.HTTP2, http2);
 

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/impl/ClientBuilderImpl.java
@@ -82,7 +82,7 @@ public class ClientBuilderImpl extends ClientBuilder {
     private ClientLogger clientLogger = new DefaultClientLogger();
     private String userAgent = RestClientRequestContext.DEFAULT_USER_AGENT_VALUE;
 
-    private boolean enableCompression;
+    private Boolean enableCompression;
 
     public ClientBuilderImpl() {
         configuration = new ConfigurationImpl(RuntimeType.CLIENT);
@@ -202,8 +202,8 @@ public class ClientBuilderImpl extends ClientBuilder {
         return this;
     }
 
-    public ClientBuilder enableCompression() {
-        this.enableCompression = true;
+    public ClientBuilder enableCompression(boolean enableCompression) {
+        this.enableCompression = enableCompression;
         return this;
     }
 
@@ -278,7 +278,7 @@ public class ClientBuilderImpl extends ClientBuilder {
             }
         }
 
-        if (enableCompression) {
+        if (Boolean.TRUE.equals(enableCompression)) {
             configuration.register(ClientGZIPDecodingInterceptor.class);
         }
 

